File: src/windows_navigator/filter.py
"""Filter and sort window list by a search query."""

from __future__ import annotations

from windows_navigator.models import WindowInfo


def _tokens_match(w: WindowInfo, tokens: list[str]) -> bool:
    """True if every token in *tokens* appears in title or process name."""
    title = w.title.casefold()
    proc = w.process_name.casefold()
    return all(t in title or t in proc for t in tokens)


def filter_windows(
    windows: list[WindowInfo],
    query: str,
    desktop_nums: set[int] | None = None,
) -> list[WindowInfo]:
    """Return windows matching *query* and optional *desktop_nums*.

    *desktop_nums* restricts results to windows on any of those desktops (OR logic);
    ``None`` means no desktop restriction.  *query* is matched case-insensitively
    against title and process name (all whitespace-separated tokens must match).
    An empty query with no desktop filter returns all windows unchanged.
    """
    result = list(windows)
    if desktop_nums:
        result = [w for w in result if w.desktop_number in desktop_nums]
    tokens = query.casefold().split()
    if not tokens:
        return result
    return [w for w in result if _tokens_match(w, tokens)]

File: go/internal/navigator/filter.go
package navigator

import "strings"

// tokensMatch returns true if every whitespace-separated token in query appears
// (case-insensitively) in w's title or process name.
func tokensMatch(w WindowInfo, query string) bool {
	title := strings.ToLower(w.Title)
	proc := strings.ToLower(w.ProcessName)
	for _, tok := range strings.Fields(strings.ToLower(query)) {
		if !strings.Contains(title, tok) && !strings.Contains(proc, tok) {
			return false
		}
	}
	return true
}

// FilterWindows returns windows matching query and desktopNums.
// desktopNums restricts to those desktops (OR logic); nil means no desktop filter.
// query is matched case-insensitively against title and process name (multi-token AND).
// An empty query with nil desktopNums returns a copy of all windows.
func FilterWindows(windows []WindowInfo, query string, desktopNums map[int]struct{}) []WindowInfo {
	hasDesktopFilter := len(desktopNums) > 0
	hasTextFilter := len(strings.Fields(query)) > 0

	if !hasDesktopFilter && !hasTextFilter {
		out := make([]WindowInfo, len(windows))
		copy(out, windows)
		return out
	}

	var result []WindowInfo
	for _, w := range windows {
		if hasDesktopFilter {
			if _, ok := desktopNums[w.DesktopNumber]; !ok {
				continue
			}
		}
		if hasTextFilter && !tokensMatch(w, query) {
			continue
		}
		result = append(result, w)
	}
	return result
}
